### ISSUE TITLE
Redis/server high CPU usage

### ISSUE THREAD

**Issue by MattHalloran** [Vrooli/Vrooli#101]
### Describe the bug
The production server is sometimes slow. This is because of some issue with Redis, which causes high CPU usage for the redis and server Docker containers. Restarting the redis container fixes the issue temporarily. I have not been able to reproduce this on the test server, which is strange.

### To Reproduce
1. Deploy on VPS
2. Click around the site for a few minutes
3. Check CPU usage with `docker stats`

## Logs
Here are logs from the redis container, which may help identify the issue:  
```
1:S 07 Aug 2022 15:19:21.672 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:21.797 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:21.921 # Failed to read response from the server: Connection reset by peer
1:S 07 Aug 2022 15:19:21.921 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:22.675 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:22.675 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:22.795 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:22.916 # Failed to read response from the server: Connection reset by peer
1:S 07 Aug 2022 15:19:22.916 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:23.678 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:23.678 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:23.798 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:23.919 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:23.919 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:24.681 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:24.681 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:24.805 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:24.930 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:24.930 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:25.684 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:25.684 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:25.811 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:25.939 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:25.939 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:26.687 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:26.687 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:26.807 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:26.927 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:26.927 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:27.689 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:27.690 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:27.809 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:27.928 # Failed to read response from the server: Connection reset by peer
1:S 07 Aug 2022 15:19:27.928 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:28.692 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:28.693 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:28.817 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:28.942 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:28.943 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:29.695 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:29.695 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:29.815 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:29.937 * Master replied to PING, replication can continue...
1:S 07 Aug 2022 15:19:30.177 * Partial resynchronization not possible (no cached master)
1:S 07 Aug 2022 15:19:30.298 * Full resync from master: ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ:1
1:S 07 Aug 2022 15:19:30.298 * MASTER <-> REPLICA sync: receiving 54976 bytes from master to disk
1:S 07 Aug 2022 15:19:30.541 * MASTER <-> REPLICA sync: Flushing old data
1:S 07 Aug 2022 15:19:30.541 * MASTER <-> REPLICA sync: Loading DB in memory
1:S 07 Aug 2022 15:19:30.547 # Wrong signature trying to load DB from file
1:S 07 Aug 2022 15:19:30.547 # Failed trying to load the MASTER synchronization DB from disk: Invalid argument
1:S 07 Aug 2022 15:19:30.547 * Reconnecting to MASTER 45.153.231.17:8886 after failure
1:S 07 Aug 2022 15:19:30.547 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:30.667 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:30.786 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:30.786 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:31.703 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:31.703 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:31.830 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:31.958 # Failed to read response from the server: Connection reset by peer
1:S 07 Aug 2022 15:19:31.958 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:32.706 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:32.706 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:32.837 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:32.973 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:32.973 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:33.710 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:33.713 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:33.839 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:33.964 # Failed to read response from the server: Connection reset by peer
1:S 07 Aug 2022 15:19:33.964 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:34.716 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:34.716 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:34.836 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:34.956 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:34.956 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:35.720 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:35.721 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:35.840 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:35.959 # Failed to read response from the server: Connection reset by peer
1:S 07 Aug 2022 15:19:35.959 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:36.724 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:36.724 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:36.843 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:36.965 # Failed to read response from the server: Connection reset by peer
1:S 07 Aug 2022 15:19:36.965 # Master did not respond to command during SYNC handshake
1:S 07 Aug 2022 15:19:37.727 * Connecting to MASTER 45.153.231.17:8886
1:S 07 Aug 2022 15:19:37.727 * MASTER <-> REPLICA sync started
1:S 07 Aug 2022 15:19:37.847 * Non blocking connect for SYNC fired the event.
1:S 07 Aug 2022 15:19:37.967 # Failed to read response from the server: Invalid argument
1:S 07 Aug 2022 15:19:37.967 # Master did not respond to command during SYNC handshake
1:signal-handler (1659885578) Received SIGTERM scheduling shutdown...
1:S 07 Aug 2022 15:19:38.530 # User requested shutdown...
1:S 07 Aug 2022 15:19:38.530 * Saving the final RDB snapshot before exiting.
1:S 07 Aug 2022 15:19:38.534 * DB saved on disk
1:S 07 Aug 2022 15:19:38.534 # Redis is now ready to exit, bye bye...
1:C 07 Aug 2022 15:19:39.453 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:C 07 Aug 2022 15:19:39.453 # Redis version=7.0.4, bits=64, commit=00000000, modified=0, pid=1, just started
1:C 07 Aug 2022 15:19:39.453 # Configuration loaded
1:M 07 Aug 2022 15:19:39.453 * monotonic clock: POSIX clock_gettime
1:M 07 Aug 2022 15:19:39.460 * Running mode=standalone, port=6379.
1:M 07 Aug 2022 15:19:39.460 # Server initialized
1:M 07 Aug 2022 15:19:39.464 * Reading RDB base file on AOF loading...
1:M 07 Aug 2022 15:19:39.464 * Loading RDB produced by version 7.0.4
1:M 07 Aug 2022 15:19:39.464 * RDB age 8415 seconds
1:M 07 Aug 2022 15:19:39.464 * RDB memory usage when created 1.39 Mb
1:M 07 Aug 2022 15:19:39.464 * RDB is base AOF
1:M 07 Aug 2022 15:19:39.464 * Done loading RDB, keys loaded: 0, keys expired: 0.
1:M 07 Aug 2022 15:19:39.465 * DB loaded from base file appendonly.aof.92.base.rdb: 0.001 seconds
1:M 07 Aug 2022 15:19:39.465 * DB loaded from append only file: 0.001 seconds
1:M 07 Aug 2022 15:19:39.465 * Opening AOF incr file appendonly.aof.92.incr.aof on server start
1:M 07 Aug 2022 15:19:39.465 * Ready to accept connections
```

**Comment by henriquenery1** [comment]
https://github.com/redis/redis/issues/7506

**Comment by MattHalloran** [comment]
@henriquenery1 Thanks for linking that issue! While it didn't solve the problem Vrooli was running into, it introduced me to health checks, which I've now added.

Luckily, while figuring out how to secure redis, I believe [the problem has been fixed](https://github.com/Vrooli/Vrooli/commit/0955743d347974ca11d3d66874e892a963508376). I'm not sure why this security fix would also solve the high CPU issue, but I've been running this fix on a test server for over a week now and the issue hasn't come up. 

I'm going to close this issue for now, but I will reopen it if the problem comes up again
